### PR TITLE
Create node implementation

### PR DESCRIPTION
**Pull request by virbas**
please review and merge into develop.

+ create node implementation
+ unit tests

### CODE REVIEW

**Comment by virbas** [comment]
test were fixed

File: src/app/common/models/node.model.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { FieldMapFromServer, ReleaseMicroschemaInfoFromServer, NodeResponse } from './server-models';
+import { FieldMapFromServer, ReleaseMicroschemaInfoFromServer, NodeResponse, NodeCreateRequest } from './server-models';
```
Comment (michaelbromley) [inline src/app/common/models/node.model.ts]: Unused import, remove this.

File: package.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
     "start-original": "ng serve",
     "start": "ng serve --verbose --proxy-config proxy.conf.json",
     "build": "ng build --prod",
-    "test": "ng test",
+    "test": "ng test -sm=false",
```
Comment (michaelbromley) [inline package.json]: Try running tests without this flag and see if the error still persists. If no, remove the flag.

File: src/app/core/providers/effects/list-effects.service.ts [hunk 3]
```diff
@@ -44,15 +44,19 @@ export class ListEffectsService {
...
                 this.state.actions.list.fetchNodeSuccess(response);
             }, error => {
                 this.state.actions.list.fetchChildrenError();
                 throw new Error('TODO: Error handling');
             });
 
-        // Refresh child node list
+       this.loadChildren(projectName, containerUuid, language);
+    }
+
```
Comment (michaelbromley) [inline src/app/core/providers/effects/list-effects.service.ts]: Add docs here - document all public methods (even if just a few words)

File: src/app/core/providers/navigation/navigation.service.ts [hunk 4]
```diff
@@ -91,8 +108,15 @@ export class NavigationService {
             outlets.detail = null;
         }
         if (instruction.detail) {
-            const { projectName, nodeUuid, language } = instruction.detail;
-            outlets.detail = [projectName, nodeUuid, language];
+            const { projectName, nodeUuid, schemaUuid, parentNodeUuid, language, command} = instruction.detail;
```
Comment (michaelbromley) [inline src/app/core/providers/navigation/navigation.service.ts]: add space after `command`

File: src/app/editor/components/create-node-button/create-node-button.component.ts [hunk 5]
```diff
@@ -1,13 +1,17 @@
 import { ChangeDetectionStrategy, Component } from '@angular/core';
 import { Observable } from 'rxjs/Observable';
 
+import { ApplicationStateService } from '../../../state/providers/application-state.service';
+import { NavigationService } from '../../../core/providers/navigation/navigation.service';
 import { EntitiesService } from '../../../state/providers/entities.service';
 import { Schema } from '../../../common/models/schema.model';
+import { Subscription } from 'rxjs/Subscription';
```
Comment (michaelbromley) [inline src/app/editor/components/create-node-button/create-node-button.component.ts]: Move this import to line 3

File: src/app/editor/components/create-node-button/create-node-button.component.ts [hunk 6]
```diff
@@ -19,7 +23,10 @@ export interface SchemaDisplayProperties {
 export class CreateNodeButtonComponent {
     schemas$: Observable<SchemaDisplayProperties[]>;
 
-    constructor(private entities: EntitiesService) {
+    constructor(
```
Comment (michaelbromley) [inline src/app/editor/components/create-node-button/create-node-button.component.ts]: get rid of the newline after the `(`

File: src/app/editor/components/create-node-button/create-node-button.component.ts [hunk 7]
```diff
@@ -29,8 +36,21 @@ export class CreateNodeButtonComponent {
             );
     }
 
-    itemClick(): void {
-        console.warn('Not implemented');
+    itemClick(schema: SchemaDisplayProperties): void {
+        let askUserToSave: Promise<boolean>;
```
Comment (michaelbromley) [inline src/app/editor/components/create-node-button/create-node-button.component.ts]: This logic might be redundant due to the canDeactivate guard on the detail outlet. We should check this together.

File: src/app/editor/components/node-editor/node-editor.component.ts [hunk 8]
```diff
@@ -1,14 +1,19 @@
...
 import { ApplicationStateService } from '../../../state/providers/application-state.service';
 import { Observable } from 'rxjs/Observable';
 import { FormGeneratorComponent } from '../../form-generator/components/form-generator/form-generator.component';
 import { EntitiesService } from '../../../state/providers/entities.service';
 import { simpleCloneDeep } from '../../../common/util/util';
+import { initializeNode } from '../../form-generator/common/initialize-node';
+import { NodeReferenceFromServer, NodeResponse } from '../../../common/models/server-models';
+import { I18nService } from '../../../core/providers/i18n/i18n.service';
+import { ListEffectsService } from '../../../core/providers/effects/list-effects.service';
+import { debuglog } from 'util';
```
Comment (michaelbromley) [inline src/app/editor/components/node-editor/node-editor.component.ts]: probably shouldn't be there

File: src/app/editor/components/node-editor/node-editor.component.ts [hunk 9]
```diff
@@ -20,43 +25,68 @@ import { simpleCloneDeep } from '../../../common/util/util';
...
-            this.state.select(state => state.editor.openNode && state.editor.openNode.uuid),
-            this.state.select(state => state.editor.openNode && state.editor.openNode.language)
-        )
-            .switchMap(([uuid, language]) => {
-                const node$ = this.entities.selectNode(uuid, { language });
-                return Observable.combineLatest(
-                    node$.filter<MeshNode>(Boolean).map(node => simpleCloneDeep(node)),
-                    node$.switchMap(node => this.entities.selectSchema(node.schema.uuid))
-                );
+        const openNode$ = this.state.select(state => state.editor.openNode)
```
Comment (michaelbromley) [inline src/app/editor/components/node-editor/node-editor.component.ts]: make sure to clean up this subscription in ngOnDestroy()

File: src/app/editor/components/node-editor/node-editor.component.ts [hunk 10]
```diff
@@ -68,19 +98,29 @@ export class NodeEditorComponent implements OnInit, OnDestroy {
...
-        return [this.node.project.name, ...breadcrumbs].join(' › ');
+        const parentNode = this.entities.getNode(this.node.parentNode.uuid, { language : this.node.language });
+        const parentDisplayNode: any = { displayName: parentNode.displayName || '' };
+        let breadcrumb = this.node.breadcrumb;
+
+        if (!breadcrumb && parentNode) {
+            const createNodeBreadcrumb: any = { displayName: this.i18n.translate('editor.create_node') };
+            breadcrumb = [createNodeBreadcrumb, parentDisplayNode, ...parentNode.breadcrumb];
+        } else if (!breadcrumb) {
+            breadcrumb = [parentDisplayNode as any, ...parentNode.breadcrumb];
```
Comment (michaelbromley) [inline src/app/editor/components/node-editor/node-editor.component.ts]: Let's take a look to see if we can do the typing better here.

File: src/app/editor/providers/editor-effects.service.ts [hunk 11]
```diff
@@ -33,6 +33,53 @@ export class EditorEffectsService {
             });
     }
 
+    createNode(projectName: string, schemaUuid: string, parentNodeUuid: string, language: string): void {
```
Comment (michaelbromley) [inline src/app/editor/providers/editor-effects.service.ts]: Add docs

File: src/app/editor/providers/editor-effects.service.ts [hunk 12]
```diff
@@ -33,6 +33,53 @@ export class EditorEffectsService {
...
+            .subscribe(response => {
+                this.state.actions.list.fetchNodeSuccess(response);
+                this.state.actions.editor.openNewNode(projectName, schemaUuid, parentNodeUuid, language);
+            }, error => {
+                this.state.actions.list.fetchChildrenError();
+                throw new Error('TODO: Error handling');
+            });
+    }
+
+    saveNewNode(projectName: string, node: MeshNode): Promise<MeshNode | void> {
```
Comment (michaelbromley) [inline src/app/editor/providers/editor-effects.service.ts]: Add docs

File: src/app/editor/providers/editor-effects.service.ts [hunk 13]
```diff
@@ -33,6 +33,53 @@ export class EditorEffectsService {
...
+        // TODO: save the new new node to the Mesh
+        console.log('Saving node', node);
```
Comment (michaelbromley) [inline src/app/editor/providers/editor-effects.service.ts]: remove

File: src/app/editor/providers/editor-effects.service.ts [hunk 14]
```diff
@@ -33,6 +33,53 @@ export class EditorEffectsService {
...
+            fields: node.fields,
+            parentNode: node.parentNode,
+            schema: node.schema,
+            language: node.language || this.config.FALLBACK_LANGUAGE,
+        };
+
+        const language = node.language || this.config.FALLBACK_LANGUAGE;
+
+        return this.api.project.createNode({ project: projectName }, nodeCreateRequest)
+        .toPromise()
```
Comment (michaelbromley) [inline src/app/editor/providers/editor-effects.service.ts]: indent operators by 1 indentation

**Pull request merged**